packages/kmod-6.1-nvidia/nvidia-open-gpu-modules.service.in

```diff
@@ -0,0 +1,22 @@
+[Unit]
+Description=Copy Open GPU drivers to needed modules directory
+Before=load-kernel-modules.service
+After=systemd-tmpfiles-setup.service
+DefaultDependencies=no
+RefuseManualStart=true
+RefuseManualStop=true
+ConditionPathExists=/run/nvidia/nvidia-open-gpu
+
+[Service]
+Type=oneshot
+# This ensures the directory is created before copying. The directory is created by tmpfiles but udev kicks this off early in boot
+ExecStart=/usr/bin/mkdir -p __NVIDIA_MODULES__
+# This copies the files from the factory directory which won't be found by modprobe to the proper driver directory which driverdog 
+# will expect. This prevents auto-loading of the drivers when teh proprietary driver is required.
+ExecStart=/usr/bin/cp -r -n __NVIDIA_MODULES__ __PREFIX__/lib/modules/__KERNEL_VERSION__/kernel/drivers/extra/video/nvidia/open-gpu
+
+RemainAfterExit=true
+StandardError=journal+console
+
+[Install]
+WantedBy=load-kernel-modules.service
```

Comment:
This is very racy - `systemd` will always queue the job immediately, and if it happens to start before `ghostdog` finishes creating the link, then the condition check won't pass and it will fail to start.

It looks like you're working around that by pushing the job after `systemd-tmpfiles-setup.service`, but that's a hack.

Zooming out a bit, it seems like you've created two synchronization problems for yourself:
* the need for the correct config file to be in place before either `driverdog link` or `driverdog load` can run
* the need to copy the .ko files into place before `driverdog load` runs, in the open GPU case

You can eliminate the second coordination problem here by making it a `driverdog` config problem instead - having a `copy-source` field instead of `link-objects` to tell `driverdog` where to find the `.ko`.

I know I pushed back on more `driverdog` functionality here but I can be wrong sometimes. It would let you delete this unit (and its sync problem) and focus only on the "correct config for driverdog" problem.

Reply:
I've resolved this by moving away from `udev` entirely. We can determine at the time of the unit file running what decision to make. It costs _very little_ to do the `ghostdog match-nvidia-driver` for these invocations so keeping track of these decisions somewhere else don't look to be worth it. Testing on a g3.4xlarge which should have to crawl the entire file and not make a match (worst case runtime) is still insignificant compared to the time it takes to link/load the modules:

```
bash-5.1# time ghostdog match-nvidia-driver tesla

real    0m0.005s
user    0m0.002s
sys     0m0.003s
bash-5.1# time ghostdog match-nvidia-driver open-gpu
open-gpu is not preferred driver: tesla

real    0m0.005s
user    0m0.005s
sys     0m0.000s
```
I can be more precise in the measurement if needed, but it looks to be that `ghostdog` is very fast at providing this decision.